### PR TITLE
Allow `#warning` and `#error` directives in case position of a switch

### LINKED ISSUE #3251: [parser]: #warning inside switch statement causes "error: new Swift parser generated errors for code that C++ parser accepted"
### Description

This code produces the error
`error: new Swift parser generated errors for code that C++ parser accepted`

### Reproduction

```swift
switch x {
            #warning("Something")
        case "a": 
…
}
```

### Expected behavior

Both parsers should allow this

### Environment

compiling using aarch64-swift-linux-musl static build for linux on macOS

### Additional information

_No response_

### PR DESCRIPTION
**Pull request by nenadvulic**
Fixes #3251.

The C++ parser accepts `#warning` and `#error` between switch cases (`parseStmtCases` has a dedicated `pound_warning`/`pound_error` branch), but SwiftParser sent them through the missing-case-label recovery path. As a result, swiftc reported `error: new Swift parser generated errors for code that C++ parser accepted` for code like:

```swift
switch x {
#warning("Something")
case "a":
  break
}
```

Changes:
- Add `macroExpansionDecl` to the element choices of `SwitchCaseListSyntax`, since `#warning`/`#error` are represented as `MacroExpansionDeclSyntax` (there is no dedicated pound-diagnostic node).
- Parse `#warning`/`#error` in `parseSwitchCases` as `macroExpansionDecl` elements. Only these two directives are accepted, mirroring the C++ parser; other macro expansions in case position still go through case-label recovery.
- Add a 604 release note since the new `SwitchCaseListSyntax.Element` case is source-breaking for exhaustive switches.

### CODE REVIEW

**Comment by nenadvulic** [comment]
a little reminder about this PR @ahoppen thanks 😊

File: Release Notes/604.md [hunk 1]
```diff
@@ -13,6 +13,11 @@
 
 ## API-Incompatible Changes
 
+- `SwitchCaseListSyntax.Element` gained a new case `macroExpansionDecl`.
```
Comment (hamishknight) [inline Release Notes/604.md]: I'm not sure it's worth pulling this into 6.4 at this point, can you create a `605.md` and put this in there?

File: Sources/SwiftSyntax/generated/raw/RawSyntaxNodesQRS.swift [hunk 2]
```diff
@@ -2301,9 +2301,11 @@ public struct RawSwitchCaseListSyntax: RawSyntaxNodeProtocol {
     /// ```
     case switchCase(RawSwitchCaseSyntax)
     case ifConfigDecl(RawIfConfigDeclSyntax)
+    /// The expansion of a freestanding macro in a position that expects a declaration.
+    case macroExpansionDecl(RawMacroExpansionDeclSyntax)
```
Comment (hamishknight) [inline Sources/SwiftSyntax/generated/raw/RawSyntaxNodesQRS.swift:2305]: Hmm this is an unfortunate doc comment given we only allow `#warning`/`#error`. Would be nice if we could customize the doc comments here, but that's a bigger change so happy to leave that for a subsequent PR

**Review by hamishknight — COMMENTED** [review]
Thank you! Can you also please follow [the RFC process](https://github.com/swiftlang/swift-syntax/blob/main/Contributor%20Documentation/RFC%20Process.md) for this change?

**Comment by nenadvulic** [comment]
@hamishknight  RFC thread on the Swift forums: https://forums.swift.org/t/rfc-allow-warning-and-error-directives-in-case-position-of-a-switch/87464

I've also moved the release note entry from `604.md` to a new `605.md` as suggested.

File: Tests/SwiftParserTest/StatementTests.swift [hunk 3]
```diff
@@ -231,6 +231,39 @@ final class StatementTests: ParserTestCase {
     )
   }
 
+  func testSwitchWithPoundDiagnostic() {
```
Comment (hamishknight) [inline Tests/SwiftParserTest/StatementTests.swift:234]: The post on the Swift forums reminded me, can you add a test case that also exercises `#error`/`#warning` in a `#if` within the switch?

Comment (nenadvulic) [reply Tests/SwiftParserTest/StatementTests.swift:234]: Added `testSwitchWithPoundDiagnosticInIfConfig` exercising `#warning`/`#error` inside an `#if`/`#else` within the switch.

File: Tests/SwiftParserTest/StatementTests.swift [hunk 4]
```diff
@@ -231,6 +231,99 @@ final class StatementTests: ParserTestCase {
...
+      switch x {
+      #error("Something")
+      default:
+        break
+      }
+      """
+    )
+  }
+
+  func testSwitchWithPoundDiagnosticInIfConfig() {
```
Comment (allevato) [inline Tests/SwiftParserTest/StatementTests.swift:267]: These tests are only showing what happens before the first `case`. Let's cover a couple other situations:

I'm assuming that these still parse the `#warning` as a statement of the case body and not as a new case?

```swift
switch x {
case 1:
  print()
#warning("something")
}

switch x {
case 1:
  print()
#warning("something")
case 2:
  break
}
```

There's also some fairly involved lookahead when `#if` is inside a case body to determine whether it's surrounding regular statements or another case list. What happens with these?

```swift
switch x {
case 1:
  print()
#if COND
case 2: break
#else
#warning("Something")
#endif
}

switch x {
case 1:
  print()
#if COND
#warning("Something")
#else
case 2: break
#endif
}

switch x {
case 1:
  print()
#if COND
#if COND2
#warning("Something")
#endif
#else
#if COND2
case 2: break
#endif
#endif
}
```

My expectation is that those would all parse correctly as `IfConfig`s wrapping `SwitchCaseList`s, not as regular statements inside the body. (Although, I have no idea what the C++ parser does here, either.)

Comment (nenadvulic) [reply Tests/SwiftParserTest/StatementTests.swift:267]: @allevato I tested all three against swiftc (Apple Swift 6.3) and SwiftSyntax already
matches the compiler here. The disambiguation is on the first branch: if it
starts with `case`/`default` the `#if` is case-list level, otherwise it's
statement level and a `case` in a later branch is an error.

Example 1 (first branch is a `case`) compiles fine, only the `#warning` fires:

```swift
switch x {
case "1": print()
#if COND
case "2": break
#else
#warning("Something")
#endif
default: break
}
```

Example 2 (first branch is `#warning`, `case` in the `#else`) is rejected by
the compiler:

```swift
switch x {
case "1": print()
#if COND
#warning("Something")
#else
case "2": break
#endif
default: break
}
```

```
error: 'case' label can only appear inside a 'switch' statement
```

Note this fires even with `-D COND` (so with the `#else` inactive), so it's
structural, not a fold-time thing. SwiftSyntax emits the equivalent diagnostic.
Example 3 behaves the same way.

For the ambiguous case (a `#if` with only `#warning`/`#error`, no case label in
any branch) the compiler attaches it to the preceding case's body, same as
SwiftSyntax does now.

So I think the current behavior is correct: matching it to accept examples 2/3
would diverge from the compiler. I can add tests pinning this down if you want.
Or if there's a compiler version where these parse differently, point me at it
and I'll match that instead.

Comment (allevato) [reply Tests/SwiftParserTest/StatementTests.swift:267]: Ok, my expectations were incorrect then, but the reasoning makes sense why it does what it does. Thanks for checking!

If we don't have any existing tests covering this situation, adding some seems like a good idea just to avoid future regressions.

One more edge case that just occurred to me:

```swift
switch x {
case 1: break
#if COND
case 2: break
#else
case 3: break
#endif
#warning("Something")
default: break
```

Does the C++ parser allow this (with or without the `default: break`)? I think swift-syntax misparses it if C++ allows it (you get a malformed `SwitchCaseItem` with the `#warning` in a `CodeBlockItem`, where I think I would expect `#warning` to be parsed as a `MacroExpansionDecl` as a sibling to the `IfConfigDecl`.

Comment (nenadvulic) [reply Tests/SwiftParserTest/StatementTests.swift:267]: Checked the new one too. swiftc accepts it (only the `#warning` fires, no
errors), and swift-syntax actually parses it the way you'd expect: the
`#warning` comes out as a `MacroExpansionDecl` sibling to the `IfConfigDecl` in
the case list, not as a malformed `SwitchCaseItem`. Tree for

```swift
switch x {
case 1: break
#if COND
case 2: break
#else
case 3: break
#endif
#warning("Something")
default: break
}
```

is `[SwitchCase(case 1), IfConfigDecl(case 2 / case 3), MacroExpansionDecl(#warning), SwitchCase(default)]`,
no diagnostics. Same with the `default` removed (you just get the usual
non-exhaustive error).

And your two simpler ones parse the `#warning` as a statement in the case body,
like you assumed. So no misparse here.

I'll add tests for all of these so it doesn't regress.

**Review by hamishknight — APPROVED** [review]

**Comment by hamishknight** [comment]
@swift-ci please test

**Comment by hamishknight** [comment]
```
23:59:38  /Users/ec2-user/jenkins/workspace/swift-syntax-PR-macOS/branch-main/swift/lib/ASTGen/Sources/ASTGen/Stmts.swift:562:7: error: switch must be exhaustive
23:59:38  560 |     var allBridgedCases: [BridgedCaseStmt] = []
23:59:38  561 |     visitIfConfigElements(node, of: SwitchCaseSyntax.self) { element in
23:59:38  562 |       switch element {
23:59:38      |       |- error: switch must be exhaustive
23:59:38      |       `- note: add missing case: '.macroExpansionDecl(_)'
23:59:38  563 |       case .ifConfigDecl(let ifConfigDecl):
23:59:38  564 |         return .ifConfigDecl(ifConfigDecl)
```

@nenadvulic Can you open a PR for https://github.com/swiftlang/swift that fixes the missing case? We can cross-repo test the changes together

**Comment by nenadvulic** [comment]
> ```
> 23:59:38  /Users/ec2-user/jenkins/workspace/swift-syntax-PR-macOS/branch-main/swift/lib/ASTGen/Sources/ASTGen/Stmts.swift:562:7: error: switch must be exhaustive
> 23:59:38  560 |     var allBridgedCases: [BridgedCaseStmt] = []
> 23:59:38  561 |     visitIfConfigElements(node, of: SwitchCaseSyntax.self) { element in
> 23:59:38  562 |       switch element {
> 23:59:38      |       |- error: switch must be exhaustive
> 23:59:38      |       `- note: add missing case: '.macroExpansionDecl(_)'
> 23:59:38  563 |       case .ifConfigDecl(let ifConfigDecl):
> 23:59:38  564 |         return .ifConfigDecl(ifConfigDecl)
> ```
> 
> @nenadvulic Can you open a PR for https://github.com/swiftlang/swift that fixes the missing case? We can cross-repo test the changes together

@hamishknight  Sure, I'll open a PR on swiftlang/swift and link it here so we can cross-repo test the two together.

**Comment by nenadvulic** [comment]
@hamishknight  I opened the companion PR: https://github.com/swiftlang/swift/pull/90074
It handles the new `.macroExpansionDecl` case in ASTGen (evaluates the `#warning`/`#error` and skips it). 
It only builds against this branch, so it needs to be cross-repo tested with it.

**Comment by hamishknight** [comment]
https://github.com/swiftlang/swift/pull/90074

@swift-ci please test

**Comment by hamishknight** [comment]
https://github.com/swiftlang/swift/pull/90074

@swift-ci please test

**Comment by nenadvulic** [comment]
@hamishknight  The cross-repo smoke test there passes (macOS + Linux). This PR's standalone macOS/Linux CI only fails because it builds against compiler `main`, which doesn't have that fix yet, so the two need to be tested/merged together.

**Comment by hamishknight** [comment]
@nenadvulic I did do a cross-repo test for this PR. Seems like the recent update-checkout changes have broken things here, let me investigate

**Comment by hamishknight** [comment]
https://github.com/swiftlang/swift/pull/90074

@swift-ci please test
